Update readme.md (#3)
wawrzynek-eq/arch-kata: readme.md

# Knowledge Out of Range Exception - O'Reilly Architectural Katas: Winter 2025

## 👥 Team

We are **Knowledge Out of Range Exception**, a group of passionate professionals from Endava Poland. Our team is consists of the following members:

- Paweł Wawrzynek
- Arkadiusz Jastrzębski
- Krzysztof Szczęsny

## 🎥 Video Presentation

**Embedded video here.**

## 📑 Table of Content

- [🤔 Problem Description](#-problem-description)
- [📊 Feasibility Analysis](#-feasibility-analysis)
- [📐 Architecture and Design](#-architecture-and-design)
- [✍️ Architecture Decision Records](#️-architecture-decision-records)

## 🤔 Problem Description

Certifiable, Inc. is facing a significant increase in demand for software architecture certifications due to new laws requiring such certifications and expansion to new markets. The company currently employs 300 expert software architects, with 5 designated experts authorized to modify certification tests. They handle around 200 candidates per week in the U.S., a number expected to grow 5-10 times with overseas expansion. The certification exam costs $800, a price fixed by the Software Architecture Licensing Board (SALB).

> To manage this surge, the company consider redesign of their flagship system, SoftArchCert by incorporating Generative AI.

Reviewer: ```suggestion
To manage this surge, the company is considering a redesign of their flagship system, SoftArchCert by incorporating Generative AI.
```

# Knowledge Out of Range Exception - O'Reilly Architectural Katas: Winter 2025

## 👥 Team

> We are **Knowledge Out of Range Exception**, a group of passionate professionals from Endava Poland. Our team is consists of the following members:

Reviewer 2: ```suggestion
We are **Knowledge Out of Range Exception**, a group of passionate professionals from Endava Poland. Our team consists of the following members:
```

- Paweł Wawrzynek
- Arkadiusz Jastrzębski
- Krzysztof Szczęsny

## 🎥 Video Presentation

**Embedded video here.**

## 📑 Table of Content

- [🤔 Problem Description](#-problem-description)
- [📊 Feasibility Analysis](#-feasibility-analysis)
- [📐 Architecture and Design](#-architecture-and-design)
- [✍️ Architecture Decision Records](#️-architecture-decision-records)

## 🤔 Problem Description

> Certifiable, Inc. is facing a significant increase in demand for software architecture certifications due to new laws requiring such certifications and expansion to new markets. The company currently employs 300 expert software architects, with 5 designated experts authorized to modify certification tests. They handle around 200 candidates per week in the U.S., a number expected to grow 5-10 times with overseas expansion. The certification exam costs $800, a price fixed by the Software Architecture Licensing Board (SALB).

Reviewer 2: ```suggestion
Certifiable, Inc. is facing a significant increase in demand for software architecture certifications due to new regulations, as well as expansion to new markets. The company currently employs 300 expert software architects, with 5 designated experts authorized to modify certification tests. They handle around 200 candidates per week in the U.S., a number expected to grow 5-10 times with overseas expansion. The certification exam costs $800, a price fixed by the Software Architecture Licensing Board (SALB).
```

To manage this surge, the company consider redesign of their flagship system, SoftArchCert by incorporating Generative AI.

### Requirements

- Automate the certification process to handle increased volume.
- Maintain or improve the accuracy and reliability of the grading process.
- Reduce the workload on human graders while ensuring high standards.

## 📊 Feasibility Analysis

### Introduction

This feasibility analysis evaluates the potential benefits, challenges, and overall viability of incorporating AI into the certification process.

![Throughput](ADR/ADR0001-throughput.png "Throughput Analysis")

**Demand Growth:**

![Demand growth study](Demand-growth-study.png "Demand growth study")

1. Expected demand for candidates without Generative AI is projected to increase significantly, reaching ~2 000 application per week by 2029.

> 2. Based on throughput analysis, to meet this demand without generative AI, the company must increase headcount by 200% or maintain a workforce of 500 full-time employees. 

Reviewer 2: To moje zdanie, ale jak je teraz czytam to widzę, że jest dość niejasne. Chodziło chyba o to, że aktualnie nasi kontraktorzy pracują średnio po 8h tygodniowo. Żeby sprostać wymaganiom, musielibyśmy albo zwiększyć ilość kontraktorów o 200% albo zamienić ich wszystkich na etatowych (35h tygodniowo) i jeszcze zatrudnić ich więcej (300 => 500)
```suggestion
2. Based on throughput analysis, to meet this demand without generative AI, the company must either increase the contractors' headcount by 200%, or employ them full-time and still extend from 300 to 500.
```

See [ADR-001](./ADR/ADR%200001%20-%20Replace%20manual%20grading%20with%20AI%20assisted%20grading.md)

**Costs:**

![Cost study](Cost-study.png "Cost study")

> 1. The estimated costs for employee work hours without automation is projected to be near $1 million annually by 2029. (200% increase in headcount would triple workforce costs)

Reviewer 2: Either "estimated" or "is projected", but not both.
```suggestion
1. The cost of employee labour, without automation, is estimated for around $1 million annually by 2029. (200% increase in headcount would triple workforce costs)
```

2. **Compliance Risks**: Ensuring adherence to regulatory requirements and maintaining transparency in the AI-assisted grading process.

### Conclusion

The feasibility analysis indicates that without AI, the company will struggle to scale efficiently due to the excessive workforce and financial demands. Integrating Generative AI into the SoftArchCert system is viable and offers significant benefits in terms of scalability, efficiency, and cost savings. While there are challenges and risks, they can be mitigated through careful planning, continuous monitoring, and a balanced approach to AI and human oversight. The proposed redesign will enable Certifiable, Inc. to manage the anticipated increase in certification requests effectively and maintain high standards in the certification process.

## 📐 Architecture and Design

### Test 1 Architecture

![Test 1](./Test-1.png "Test 1")

The architecture diagram ilustrates the automated grading process for Test 1. It shows high overview of how responses are processed and graded.

- Candidate complete Test 1, consisting of multiple-choice and short-answere questions.
- Responses are stored in a databse as ungraded answere

> - Multiple-choice questions are graded through an autograding system. **No impact on the currect process**

Reviewer: ```suggestion
- Multiple-choice questions are graded through an autograding system. **No impact on the current process**
```

- Short-answer responses are passed to an [AI Grading](#grading-component) for evaluation. **Human grading process is replaced by new AI Grading**
- System compiles all graded responses and stores them in a database
- Candidates are notified via email.

### Test 2 Architecture

![Test 2](./Test-2.png "Test 2")

> Test 2 is designed to assess candidates ability to create architectural solution. Suggested approach modify current process to evaluation based on [AI Grading](#grading-component)

Reviewer: ```suggestion
Test 2 is designed to assess candidates ability to create an architectural solution. Suggested approach modifies the current process by adding evaluation based on [AI Grading](#grading-component)
```

> - Candidate complete Test 2 by designing an architectural solution

Reviewer: ```suggestion
- Candidate completes Test 2 by designing an architectural solution
```

- Solution is uploaded to the Architectural Submissions Database
- AI analyzes the submission using predefined evaluation criteria such as design quality, correctness, etc.
- AI assigns an initial grade and feedback 

> - Expert Review submission. Hybrid approach ensures quality control while minimizing manual effort.

Reviewer: ```suggestion
- Expert reviews the submission. Hybrid approach ensures quality control while minimizing manual effort.
```

- Test 2 grading results are stored
- If candidate passes, their certification status is updated in the Certification Database

> - Candidate receive their final result via email.

Reviewer: ```suggestion
- Candidate receives his final result via email.
```

### Grading Component Architecture

![Grading Component](./Grading-component.png "Grading Component")

> The AI-Assisted Grading is designed to automate and enchance the grading process by leveraging multiple AI models. This system reduce expert workload.

Reviewer: ```suggestion
The AI-Assisted Grading is designed to automate and enhance the grading process by leveraging multiple AI models. This system reduces expert workload.
```

> - Candidate answers are fed into AI Prompt Generator.

Reviewer: ```suggestion
- Candidate answers are fed into the AI Prompt Generator.
```

> - Prompt Generator formulates queries and requests grading suggestions from multiple AI Models. Several AI models independently evaluate responses.

Reviewer: ```suggestion
- Prompt Generator formulates queries and requests grading suggestions from multiple AI Models, each with multiple different prompts. Responses are evaluated independently.
```

> - Result from multiple AI models are aggregated to generate a final grading suggestion

Reviewer: ```suggestion
- Results are aggregated to generate a final grading suggestion
```

> - Expert review and adjust AI generated grades if necessary

Reviewer: ```suggestion
- Expert reviews and adjusts AI generated grades if necessary
```
